server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java

```diff
@@ -51,103 +48,191 @@
 public final class FetchDocValuesPhase implements FetchSubPhase {
 
     @Override
-    public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOException {
-
+    public FetchSubPhaseExecutor getExecutor(SearchContext context) throws IOException {
         if (context.collapse() != null) {
             // retrieve the `doc_value` associated with the collapse field
             String name = context.collapse().getFieldName();
             if (context.docValuesContext() == null) {
                 context.docValuesContext(new FetchDocValuesContext(
-                        Collections.singletonList(new FieldAndFormat(name, null))));
+                    Collections.singletonList(new FieldAndFormat(name, null))));
             } else if (context.docValuesContext().fields().stream().map(ff -> ff.field).anyMatch(name::equals) == false) {
                 context.docValuesContext().fields().add(new FieldAndFormat(name, null));
             }
         }
 
         if (context.docValuesContext() == null) {
-            return;
+            return null;
         }
 
+        List<DocValueField> fields = new ArrayList<>();
         for (FieldAndFormat fieldAndFormat : context.docValuesContext().fields()) {
-            String field = fieldAndFormat.field;
-            MappedFieldType fieldType = context.mapperService().fieldType(field);
-            if (fieldType != null) {
-                final IndexFieldData<?> indexFieldData = context.getForField(fieldType);
-                final boolean isNanosecond;
-                if (indexFieldData instanceof IndexNumericFieldData) {
-                    isNanosecond = ((IndexNumericFieldData) indexFieldData).getNumericType() == NumericType.DATE_NANOSECONDS;
-                } else {
-                    isNanosecond = false;
-                }
-                final DocValueFormat format;
-                String formatDesc = fieldAndFormat.format;
-                if (isNanosecond) {
-                    format = withNanosecondResolution(fieldType.docValueFormat(formatDesc, null));
-                } else {
-                    format = fieldType.docValueFormat(formatDesc, null);
+            DocValueField f = buildField(context, fieldAndFormat);
```

Comment:
I assume this is okay, but wanted to check -- before we looped over fields then hits, but now we loop over hits then fields. This means that we hold onto a (potentially large) number of doc values iterators for the duration of the fetch phase. Are there any concerns around this change?

Reply:
I can see that it may cause memory spikes, but I think DV iterators are generally low-impact.  I'll try and run some benchmarks to see if there's an appreciable difference